src/main/java/com/eerussianguy/blazemap/mixin/ServerLevelMixin.java

```diff
@@ -1,26 +1,44 @@
 package com.eerussianguy.blazemap.mixin;
 
+import java.util.List;
+import java.util.concurrent.Executor;
+
 import net.minecraft.core.BlockPos;
+import net.minecraft.core.Holder;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.server.MinecraftServer;
 import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.progress.ChunkProgressListener;
+import net.minecraft.world.level.Level;
 import net.minecraft.world.level.ChunkPos;
+import net.minecraft.world.level.CustomSpawner;
 import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.level.chunk.ChunkGenerator;
+import net.minecraft.world.level.dimension.DimensionType;
+import net.minecraft.world.level.storage.LevelStorageSource;
+import net.minecraft.world.level.storage.ServerLevelData;
 
 import com.eerussianguy.blazemap.engine.server.BlazeMapServerEngine;
+
 import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
 import org.spongepowered.asm.mixin.injection.At;
 import org.spongepowered.asm.mixin.injection.Inject;
 import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;
 
 @Mixin(ServerLevel.class)
-public abstract class ServerLevelMixin {
+public abstract class ServerLevelMixin extends Level {
+    // This won't be called because it's a mixin, but Java still wants it to exist at compile time,
+    // so directly copying from ServerLevel
+    public ServerLevelMixin(MinecraftServer p_203762_, Executor p_203763_, LevelStorageSource.LevelStorageAccess p_203764_, ServerLevelData p_203765_, ResourceKey<Level> p_203766_, Holder<DimensionType> p_203767_, ChunkProgressListener p_203768_, ChunkGenerator p_203769_, boolean p_203770_, long p_203771_, List<CustomSpawner> p_203772_, boolean p_203773_) {
+      super(p_203765_, p_203766_, p_203767_, p_203762_::getProfiler, false, p_203770_, p_203771_);
+    }
 
-    @Inject(method = "onBlockStateChange", at=@At("HEAD"))
-    public void onBlockStateChange(BlockPos block, BlockState s1, BlockState s2, CallbackInfo ci) {
-        // If it's stupid but it works, it's not stupid.
-        // @Shadow fucks with the project because it doesn't reobf properly.
-        // No one knows why, no one can / wants to help, and so no one can point fingers at us either.
-        ServerLevel level = (ServerLevel) (Object) this;
+    @Shadow
+    public abstract void onBlockStateChange(BlockPos block, BlockState s1, BlockState s2);
```

Comment:
Needed to `@Shadow` the function we're targetting so the Annotation Processor for the mixins would know to add it to the refmap

Reply:
Looks like we can't avoid `@Shadow` no matter what we do... but it doesn't matter since you fixed the refmap. :D